# Fine-tune LLaMA2 7b Model with PEFT method for Stock Price Prediction

reference
- https://www.kaggle.com/code/lucamassaron/fine-tune-llama-2-for-sentiment-analysis

As a first step, install the specific libraries necessary to make this work
- accelerate is a distributed traing library for PyTorch by HugglingFace. it allows you to train your models on mutiple GPU or CPUs in parallel(distributed configurations) which can significatly spped up traing in presense of multiple GPUs(I won't use it in this work.)
- peft is a python library by HuggingFace for effiecient adaptation of pre-trained language models(PLMs) to various downstream applications without fine-tuning all the model's parameters. PEFT methods only fine-tune a small number of (extra) model parameters, thereby greatly decreasing the computational and storage costs.
- bitsandbytes by Time Dettmers,is a lightweight wrapper around CUDA custom functions,in particular 8-bit optimizers,matrix multiplication(LLM.int8()), and quantization functions.It allows to run models stored in 4-bit precision: while 4-bit bitsandbytes stores weights in 4-bits, the computation still happens in 16 or 32-bit and here any combination can be chosen(float16,bfloat16,float32, and so on).
- transformers is a Python library for NLP, it provides a number of pre-trained models for NLP tasks such as text classification, question answering, and machine translation.
- trl is a full stack library by HuggingFace providing a set of tools to train transfomer language model with Reinforcement Learning, from the Supervised Fine-tuning step(SFT), Reward Modeling step(RM) to the Proximal Policy Optimization(PPO) step. 

The code imports the os module and sets two environment variables:
- CUDA_VISIBLE_DEVICES: This environment variables tells PyTorch which GPUs to use. In this case, the code is setting the environment variable to 0, which means that PyTorch will use the first GPU.
- CUDA_VISIBLE_DEVICES: This environment variable tells the HuggingFace Transfomers library whether to parallelize the tokenization process. In this case, the code is setting the environment variable to false, which means the the tokenization process will not be parallelized.

In [1]:
import os
os.environ["CUDA_VISIBLE_DEVICES"] = "0"
os.environ["TOKENIZERS_PARALLELISM"] = "false"

- The code import warnings;warnings.fiterwarnings("ignore") imports the warnings module and sets the warning filter to ignore. This means all warnings will be suppressed and will not be displayed. Actually during training there are many warnings that do not prevent the fine-tuning but can be distracting and make you wonder if you are doing the correct things.

In [2]:
import warnings
warnings.filterwarnings("ignore")
print("1")

1


In [4]:
!pip install -q -U "bitsandbytes==0.42.0"
!pip install -q -U git+https://github.com/huggingface/peft@4a1559582281fc3c9283892caea8ccef1d6f5a4f
!pip install git+https://github.com/huggingface/trl.git@7630f877f91c556d9e5a3baa4b6e2894d90ff84c
!pip install -q -U "accelerate==0.26.1"
!pip install -q -U  "transformers==4.39.1"
!pip install -q -U  "datasets==2.16.0"
!pip install tensorflow[and-cuda]
!pip3 install -q -U torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu121

  Cloning https://github.com/huggingface/trl.git (to revision 7630f877f91c556d9e5a3baa4b6e2894d90ff84c) to /tmp/pip-req-build-3b0dcelk
  Running command git clone --filter=blob:none --quiet https://github.com/huggingface/trl.git /tmp/pip-req-build-3b0dcelk
  Running command git rev-parse -q --verify 'sha^7630f877f91c556d9e5a3baa4b6e2894d90ff84c'
  Running command git fetch -q https://github.com/huggingface/trl.git 7630f877f91c556d9e5a3baa4b6e2894d90ff84c
  Running command git checkout -q 7630f877f91c556d9e5a3baa4b6e2894d90ff84c
  Resolved https://github.com/huggingface/trl.git to commit 7630f877f91c556d9e5a3baa4b6e2894d90ff84c
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 103.4/103.4 kB 4.9 MB/s eta 0:00:00
  Created wheel for trl: filename=trl-0.7.12.dev0-py3-none-any.whl size=173434 sha256=1a090c29d7999801f06cc5fd3337cad709b9ebbbf991ba78a710506db37039a5

In [5]:
import numpy as np
import pandas as pd
import os
from tqdm import tqdm
import bitsandbytes as bnb
import torch
import torch.nn as nn
import transformers
from datasets import Dataset
from peft import LoraConfig, PeftConfig
from trl import SFTTrainer
from trl import setup_chat_format
from transformers import (AutoModelForCausalLM, 
                          AutoTokenizer, 
                          BitsAndBytesConfig, 
                          TrainingArguments, 
                          pipeline, 
                          logging)
from sklearn.metrics import (accuracy_score, 
                             classification_report, 
                             confusion_matrix)
from sklearn.model_selection import train_test_split

2024-07-19 07:18:36.845090: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:485] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-07-19 07:18:36.901112: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:8454] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-07-19 07:18:36.917255: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1452] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered


In [6]:
print(f"pytorch version {torch.__version__}")

pytorch version 2.3.1+cu121


In [7]:
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
print(f"working on {device}")

working on cuda:0


### Log in Huggingface

In [8]:
from huggingface_hub import login
from kaggle_secrets import UserSecretsClient
user_secrets = UserSecretsClient()

hf_token = user_secrets.get_secret("Huggingface_token")

login(token = hf_token)


The token has not been saved to the git credentials helper. Pass `add_to_git_credential=True` in this function directly or `--add-to-git-credential` if using via `huggingface-cli` if you want to set the git credential as well.
Token is valid (permission: write).
Your token has been saved to /root/.cache/huggingface/token
Login successful


# Preparing the data and the core evaluation functioins
The code in the next cell performs the following steps:
1. Reads the input dataset from the all-data.csv file, which is a comma-separated value(CSV) file with two columns: sentiment and text.
2. Splits the dataset into training and test sets,with 300 samples in each set. The split is stratified by sentiment, so that each set contains a representative of positive,neutral, and negative sentiments.
3. Shuffles the train data in a replicable order(random_state=10)
4. Transfoms the texts contained in the train and test data into prompts to be used by LLamMa: the train prompts contains the expected answer we want to fine-tune the model-with
5. The residual examples not in train or test, for reporting purposes during during training (but it won't be used for early stopping), is treated as evaluatio  data, which is sampled with repetition in order to have a 50/50/50 sample (negative instances are very few, hence the shoud be repeated)
6. The train and eval data are wrapped by the class from HuggingFace's datasets library(backed by the Apache Arrow format)

This prepares in a single cell train_data, eval_data and test_data datasets to be used in the fine tuning.

In [9]:
filename = "../input/sentiment-analysis-for-financial-news/all-data.csv"

df = pd.read_csv(filename, 
                 names=["sentiment", "text"],
                 encoding="utf-8", encoding_errors="replace")

X_train = list()
X_test = list()
for sentiment in ["positive", "neutral", "negative"]:
    train, test  = train_test_split(df[df.sentiment==sentiment], 
                                    train_size=300,
                                    test_size=300, 
                                    random_state=42)
    X_train.append(train)
    X_test.append(test)

X_train = pd.concat(X_train).sample(frac=1, random_state=10)
X_test = pd.concat(X_test)

eval_idx = [idx for idx in df.index if idx not in list(X_train.index) + list(X_test.index)]
X_eval = df[df.index.isin(eval_idx)]
X_eval = (X_eval
          .groupby('sentiment', group_keys=False)
          .apply(lambda x: x.sample(n=50, random_state=10, replace=True)))
X_train = X_train.reset_index(drop=True)

def generate_prompt(data_point):
    return f"""
            Analyze the sentiment of the news headline enclosed in square brackets, 
            determine if it is positive, neutral, or negative, and return the answer as 
            the corresponding sentiment label "positive" or "neutral" or "negative".

            [{data_point["text"]}] = {data_point["sentiment"]}
            """.strip()

def generate_test_prompt(data_point):
    return f"""
            Analyze the sentiment of the news headline enclosed in square brackets, 
            determine if it is positive, neutral, or negative, and return the answer as 
            the corresponding sentiment label "positive" or "neutral" or "negative".

            [{data_point["text"]}] = """.strip()

X_train = pd.DataFrame(X_train.apply(generate_prompt, axis=1), 
                       columns=["text"])
X_eval = pd.DataFrame(X_eval.apply(generate_prompt, axis=1), 
                      columns=["text"])

y_true = X_test.sentiment
X_test = pd.DataFrame(X_test.apply(generate_test_prompt, axis=1), columns=["text"])

train_data = Dataset.from_pandas(X_train)
eval_data = Dataset.from_pandas(X_eval)

In [10]:
X_test

text
567   Analyze the sentiment of the news headline enc...
1752  Analyze the sentiment of the news headline enc...
995   Analyze the sentiment of the news headline enc...
601   Analyze the sentiment of the news headline enc...
568   Analyze the sentiment of the news headline enc...
...                                                 ...
4219  Analyze the sentiment of the news headline enc...
4814  Analyze the sentiment of the news headline enc...
4059  Analyze the sentiment of the news headline enc...
4720  Analyze the sentiment of the news headline enc...
4453  Analyze the sentiment of the news headline enc...

[900 rows x 1 columns]

Next part to do is creating a function to evaluate the results from the fine-tuned sentiment model. The function performs the following setps"
1. Maps the sentiment labels to a numeriacal representation, where 2 represents positive, 1 represents neutral, and 0 represents negative.
2. Calculates the accuracy of the model on the test data.
3. Generates an accuracy report for each sentiment labal.
4. Generates a classification report for the model.
5. Generates a confusion matrix for the model.

In [11]:
def evaluate(y_true, y_pred):
    labels = ['positive', 'neutral', 'negative']
    mapping = {'positive': 2, 'neutral': 1, 'none':1, 'negative': 0}
    def map_func(x):
        return mapping.get(x, 1)
    
    y_true = np.vectorize(map_func)(y_true)
    y_pred = np.vectorize(map_func)(y_pred)
    
    # Calculate accuracy
    accuracy = accuracy_score(y_true=y_true, y_pred=y_pred)
    print(f'Accuracy: {accuracy:.3f}')
    
    # Generate accuracy report
    unique_labels = set(y_true)  # Get unique labels
    
    for label in unique_labels:
        label_indices = [i for i in range(len(y_true)) 
                         if y_true[i] == label]
        label_y_true = [y_true[i] for i in label_indices]
        label_y_pred = [y_pred[i] for i in label_indices]
        accuracy = accuracy_score(label_y_true, label_y_pred)
        print(f'Accuracy for label {label}: {accuracy:.3f}')
        
    # Generate classification report
    class_report = classification_report(y_true=y_true, y_pred=y_pred)
    print('\nClassification Report:')
    print(class_report)
    
    # Generate confusion matrix
    conf_matrix = confusion_matrix(y_true=y_true, y_pred=y_pred, labels=[0, 1, 2])
    print('\nConfusion Matrix:')
    print(conf_matrix)

# Testing the model without fine-tuning

Next we need to take care of the model, which is a 7b-hf(7 billion parameters, no RLHF(Reinforcement Learning From Human Feedback), in the HuggingFace compatible format), loading from Kaggle models and quantization.

Model loading and quantization:
- First the code loads the LLaMA2 language model from the HuggingFace Hub.
- the code gets the float16 type from the torch library. This is the data type that will be used for the computations.
- Next, it creates a BitsAndBytesConfig object with the following setting:
    1. load_in_4bit: Load the model weights in 4-bit format.
    2. bnb-4bit-quant_type: Use the "nf4" quantization type. 4-bit NormalFloat(NF4),is a new data type that is information theoretically optimal for normally distributed weights.
    3. bnb_4bit_compute_dtype: Use the float16 data type for computations.
    4. bnb_4bit_use_double_quant: Do not use double quantization(reduces the average memory footprint by quantizing also the quantization constants and saves an additional 0.4 bits per parameter.)
- Then the code creates a AutoModelForCasualLM object from the pre-trained LLaMA2 language model, using the BitAndBytesConfig object for quantization.
- After that, the code disables caching for model.
- Finally the code sets the pre-training token probability to 1.

Tokenizer loading:
- First, the code loads the tokenizer for the LLaMA2 language model.
- Then it sets the padding token to be the end-of-sequnce(EOS) token.
- Finally, the code sets the padding side to be "right",which means that the inpus sequences will be padded on the right side. this is crucial for correct padding direction (this is the way with LLaMA2).

#### docs of BitsAndByteConfig(https://huggingface.co/docs/transformers/main/en/main_classes/quantization#transformers.BitsAndBytesConfig)
- load_in_4bit (bool, optional, defaults to False) — This flag is used to enable 4-bit quantization by replacing the Linear layers with FP4(4-bit floating-point)/NF4((normalized float 4) layers from bitsandbytes.
- bnb_4bit_quant_type (str, optional, defaults to "fp4") — This sets the quantization data type in the bnb.nn.Linear4Bit layers. Options are FP4 and NF4 data types which are specified by fp4 or nf4.
- bnb_4bit_compute_dtype (torch.dtype or str, optional, defaults to torch.float32) — This sets the computational type which might be different than the input type. For example, inputs might be fp32, but computation can be set to bf16 for speedups.
-bnb_4bit_use_double_quant (bool, optional, defaults to False) — This flag is used for nested quantization where the quantization constants from the first quantization are quantized again.

#### AutoModelForCausalLM
https://huggingface.co/docs/transformers/model_doc/auto#transformers.AutoModelForCausalLM

In [12]:
model_name = "../input/llama-2/pytorch/7b-hf/1"

compute_dtype = getattr(torch, "float16")

bnb_config = BitsAndBytesConfig(
    load_in_4bit=True, 
    bnb_4bit_quant_type="nf4", 
    bnb_4bit_compute_dtype=compute_dtype,
    bnb_4bit_use_double_quant=True,
)

model = AutoModelForCausalLM.from_pretrained(
    model_name,
    device_map=device,
    torch_dtype=compute_dtype,
    quantization_config=bnb_config, 
)

model.config.use_cache = False
model.config.pretraining_tp = 1

tokenizer = AutoTokenizer.from_pretrained(model_name, 
                                          trust_remote_code=True,
                                         )
tokenizer.pad_token = tokenizer.eos_token
tokenizer.padding_side = "right"

model, tokenizer = setup_chat_format(model, tokenizer)

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

In the next cell, we set a function for predicting the sentiment of a neas headline using the LLaMA2 language model. The function takes three arguments.

test: a Pandas DataFrame containing the news headlines to be predicted. model: The pre-trained LLaMA2 language model. tokenizer: The tokenizer for the LLaMA2 language model.

The function works as follows:
1. For each news headling in the test DataFrame:
    - Create a prompt for the language model, which asks it to analyze the sentiment of the news headline and return the corresponding sentiment label.
    - Use the pipeline() function from HuggingFace Transformers library to generate text from the language model, using the prompt.
    - Extract the predicted sentiment label from the generated text.
    - Append the predicted sentiment label to the y-pred list.
2. Return the y_pred list

- The pipeline() function from the HuggingFace Transformers library is used to generate text from the language model.The task argument specifies that the task is text generation. The model and tokenizer argument specify the pre-trained LLaMa2 language model and the tokenizer for the language model. The max_new_tokens argument specifies the maximum number of new tokens to generate. The teperature argument controls the randomness of the generated text. A lower temperature will produce mode predictable text, while a higher temperature will produce more creative and unexpected text.

- The if statement checks if the generated text contains the word "positive". If it does then the predicted sentiment label is "positive". Otherwise, the if statement checks if the generated text contains the word "negative". If it does, the the predicted sentiment label is "negative".  Otherwise, the if statement checks if the generated text contains the word "neutral". If it does, the the predicted sentiment label is "neutral".

In [13]:
def predict(test, model, tokenizer):
    y_pred = []
    for i in tqdm(range(len(X_test))):
        prompt = X_test.iloc[i]["text"]
        pipe = pipeline(task="text-generation", 
                        model=model, 
                        tokenizer=tokenizer, 
                        max_new_tokens = 1, 
                        temperature = 0.0,
                       )
        result = pipe(prompt)
        answer = result[0]['generated_text'].split("=")[-1]
        if "positive" in answer:
            y_pred.append("positive")
        elif "negative" in answer:
            y_pred.append("negative")
        elif "neutral" in answer:
            y_pred.append("neutral")
        else:
            y_pred.append("none")
    return y_pred

At this point, we are ready to test the LLaMA2 7b-hf model and see how it performs on this problem without any fine-tuning. This allows to get insights on the model itself and establish a baseline.

In [14]:
y_pred = predict(test,model,tokenizer)

100%|██████████| 900/900 [05:29<00:00,  2.73it/s]


In the following cell, we evaluate the results. There is little to be said, it is performing really terribly because the 7b-hf model tends to just predict a neutral sentiment and seldom it detects positive or negative sentiment.

In [16]:
evaluate(y_true,y_pred)

Accuracy: 0.373
Accuracy for label 0: 0.027
Accuracy for label 1: 0.937
Accuracy for label 2: 0.157

Classification Report:
              precision    recall  f1-score   support

           0       0.89      0.03      0.05       300
           1       0.34      0.94      0.50       300
           2       0.67      0.16      0.25       300

    accuracy                           0.37       900
   macro avg       0.63      0.37      0.27       900
weighted avg       0.63      0.37      0.27       900


Confusion Matrix:
[[  8 287   5]
 [  1 281  18]
 [  0 253  47]]


# Fine-tuning

In the next cell we set everything ready. for the fine-tuning, We configures and initializes a Simple Fine-tuning Trainer(SFTTrainer) for training a large language model using the PEFT method, which should save time as it operates on a reduced number of parameters compared to the model's overall size. The PEFT method focuses on refining a limited set of (additional) model parameters, while keeping the majority of the pre-trained LLM parameters fixed. This signifucatly reduces both computatioanl and storage expenses. Additionally, this strategy addresses the challenge of catastrophic forgetting, which often occurs during the coplete fine-tuning of LLMs/

### PEFTConfig:

The peft_config object specified the parameters for PEFT. THe following are some of most important parameters:

- lora_alpha: The learning rate for the LoRA update metrices.
- lora_dropout: The dropout probability for the LoRA updata matrices.
- r: The rank of the LoRA update matrics.
- bias: The type of bias to use. The possible values are none,additive, and learned.
- task_type: The type of task that the model is being trained for, The possible valuse are CAUSAL_LM and MASKED_LM.

### TrainingArguments:

The traing arguments object specifies the parameters for training the model. The following are some of the most important parameters:

- output_dir: The directory where the training logs and checkpoints will be saved.
- num_train_epochs: The number of epochs to train the model for.
- per_device_train_batch_size: The number of samples in each batch on each device.
- gradient_accumulation_steps: The number of batches to accumulate gradients before updating the model parameters.
- optim: The optimizer to use for training the model.
- save_steps: The number of steps after which to save a checkpoint.
- logging_steps: The number of steps after which to log the training metrics.
- learning_rate: The learning rate for the optimizer.
- weight_decay: The weight decay parameter for the optimizer.
- fp16: Whether to use 16-bit floating-point precision.
- bf16: Whether to use BFloat16 precision.
- max_grad_norm: The maximum gradient norm.
- max_steps: The maximum number of steps to train the model for.
- warmup_ratio: The proportion of the training steps to use for warming up the learning rate.
- group_by_length: Whether to group the training samples by length.
- lr_scheduler_type: The type of learning rate scheduler to use.
- report_to: The tools to report the training metrics to.
- evaluation_strategy: The strategy for evaluating the model during training. 

### SFTTrainer:

The SFTTrainer is a custom trainer class from the TRL library. It is used to train large language models (also using the PEFT method).

The SFTTrainer object is initialized with the following arguments:
- model: The model to be trained.
- train_dataset: The training dataset.
- eval_dataset: The evaluation dataset.
- peft_config: The PEFT configuration.
- dataset_text_field: The name of the text field in the dataset.
- tokenizer: The tokenizer to use.
- args: The training arguments.
- packing: Whether to pack the training samples.
- max_seq_length: The maximum sequence length.

Once the SFTTrainer object is intialized, it can be used to train the model by calling the train() method.

In [17]:
output_dir="fined_tuned_llama2_for_finance_sentiment"

peft_config = LoraConfig(
        lora_alpha=16, 
        lora_dropout=0.1,
        r=64,
        bias="none",
        target_modules="all-linear",
        task_type="CAUSAL_LM",
)

training_arguments = TrainingArguments(
    output_dir=output_dir,                    # directory to save and repository id
    num_train_epochs=3,                       # number of training epochs
    per_device_train_batch_size=1,            # batch size per device during training
    gradient_accumulation_steps=8,            # number of steps before performing a backward/update pass
    gradient_checkpointing=True,              # use gradient checkpointing to save memory
    optim="paged_adamw_32bit",
    save_steps=0,
    logging_steps=25,                         # log every 10 steps
    learning_rate=2e-4,                       # learning rate, based on QLoRA paper
    weight_decay=0.001,
    fp16=True,
    bf16=False,
    max_grad_norm=0.3,                        # max gradient norm based on QLoRA paper
    max_steps=-1,
    warmup_ratio=0.03,                        # warmup ratio based on QLoRA paper
    group_by_length=True,
    lr_scheduler_type="cosine",               # use cosine learning rate scheduler
    report_to="tensorboard",                  # report metrics to tensorboard
    evaluation_strategy="epoch"               # save checkpoint every epoch
)

trainer = SFTTrainer(
    model=model,
    args=training_arguments,
    train_dataset=train_data,
    eval_dataset=eval_data,
    peft_config=peft_config,
    dataset_text_field="text",
    tokenizer=tokenizer,
    max_seq_length=1024,
    packing=False,
    dataset_kwargs={
        "add_special_tokens": False,
        "append_concat_token": False,
    }
)

Map:   0%|          | 0/900 [00:00<?, ? examples/s]

Map:   0%|          | 0/150 [00:00<?, ? examples/s]

The following code will train the model using the trainer.train() method and then save the trained model to the trained model to the trained-model directory.

# LoRa configuration
peft_config = LoraConfig(
    r=16,
    lora_alpha=32,
    lora_dropout=0.10,
    bias='none',
    inference_mode=True,
    task_type=TaskType.SEQ_CLS,
    target_modules=['o_proj', 'v_proj'])
# Get peft
model_0 = get_peft_model(base_model_0, peft_config).to(device0) 
#Load weights
model_0.load_state_dict(torch.load(WEIGHTS_PATH), strict=False)
model_0.eval()

model_1 = get_peft_model(base_model_1, peft_config).to(device1)
model_1.load_state_dict(torch.load(WEIGHTS_PATH), strict=False)
model_1.eval()

#Trainable Parameters
model_0.print_trainable_parameters(), model_1.print_trainable_parameters()

In [18]:
trainer.train()

TrainOutput(global_step=336, training_loss=0.7177914522943043, metrics={'train_runtime': 3665.3631, 'train_samples_per_second': 0.737, 'train_steps_per_second': 0.092, 'total_flos': 1.0717884041527296e+16, 'train_loss': 0.7177914522943043, 'epoch': 2.99})

training output of 1st trial : TrainOutput(global_step=336, training_loss=0.7172041322503772, metrics={'train_runtime': 3666.0185, 'train_samples_per_second': 0.736, 'train_steps_per_second': 0.092, 'total_flos': 1.0717884041527296e+16, 'train_loss': 0.7172041322503772, 'epoch': 2.99})
TrainOutput(global_step=336, training_loss=0.7177914522943043, metrics={'train_runtime': 3665.3631, 'train_samples_per_second': 0.737, 'train_steps_per_second': 0.092, 'total_flos': 1.0717884041527296e+16, 'train_loss': 0.7177914522943043, 'epoch': 2.99})

The model and the tokenizer are saved to disk for later usage.

In [19]:
# Save trained model and tokenizer
trainer.save_model()
tokenizer.save_pretrained(output_dir)

('fined_tuned_llama2_for_finance_sentiment/tokenizer_config.json',
 'fined_tuned_llama2_for_finance_sentiment/special_tokens_map.json',
 'fined_tuned_llama2_for_finance_sentiment/tokenizer.model',
 'fined_tuned_llama2_for_finance_sentiment/added_tokens.json',
 'fined_tuned_llama2_for_finance_sentiment/tokenizer.json')

### Save fine-tuned model to Huggingface

In [20]:
# trainer.model.save_pretrained(output_dir)
trainer.model.push_to_hub(output_dir, use_temp_dir=True)


README.md:   0%|          | 0.00/27.0 [00:00<?, ?B/s]

adapter_model.safetensors:   0%|          | 0.00/1.69G [00:00<?, ?B/s]

CommitInfo(commit_url='https://huggingface.co/Journee-Kim/fined_tuned_llama2_for_finance_sentiment/commit/d4b080f5a232ff5efc9060ac7f146ea9ec3b613b', commit_message='Upload model', commit_description='', oid='d4b080f5a232ff5efc9060ac7f146ea9ec3b613b', pr_url=None, pr_revision=None, pr_num=None)

Afterwards, loading the TensorBoard extension and start TensorBoard, pointing to the logs/runs directory, which is assumed to contain the training logs and checkpoints for your model, will allow you to understand how the models fits during the training.

# Saving model to disk for later usage

At this point, in order to demonstrate how to re-utilize the model, we reload it from the disk and merge it with the original LLaMA model.

In factm when working with QLoRA, we exclusively train adapters instead of the entire model. So, when you save the model during training, you're only preserving the adapter weights, not the entire model. If you can merge the adapter weights into the model weights using the merge_and_upload method. Then, you can save the model using the model using the save_pretrained method. This will create a default model that's ready for inference tasks.

Before proceeding, we first remove the previous model and clean up the memory from various onjects we won't use anymore.

In [104]:
import gc
del [model, tokenizer, peft_config, trainer, train_data, eval_data, bnb_config, training_arguments]
del [df, X_train, X_eval]
del [TrainingArguments, SFTTrainer, LoraConfig, BitsAndBytesConfig]


NameError: name 'peft_config' is not defined

In [105]:
for _ in range(100):
    torch.cuda.empty_cache()
    gc.collect()

In [103]:
!nvidia-smi

Fri Jul 19 09:29:34 2024       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.90.07              Driver Version: 550.90.07      CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  Tesla P100-PCIE-16GB           Off |   00000000:00:04.0 Off |                    0 |
| N/A   37C    P0             32W /  250W |   14217MiB /  16384MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+-----

Then we can proceed to merging the weights and we will be using the merged model for our purposes.

In [26]:
from peft import AutoPeftModelForCausalLM

finetuned_model = "./fined_tuned_llama2_for_finance_sentiment/"
compute_dtype = getattr(torch,'float16')
tokenizer = AutoTokenizer.from_pretrained("/kaggle/input/llama-2/pytorch/7b-hf/1")

model =  AutoPeftModelForCausalLM.from_pretrained(
     finetuned_model,
     torch_dtype=compute_dtype,
     return_dict=False,
     low_cpu_mem_usage=True,
     device_map=device,
)

merged_model = model.merge_and_unload()
merged_model.save_pretrained("./merged_model",safe_serialization=True, max_shard_size="2GB")
tokenizer.save_pretrained("./merged_model")

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
Some non-default generation parameters are set in the model config. These should go into a GenerationConfig file (https://huggingface.co/docs/transformers/generation_strategies#save-a-custom-decoding-strategy-with-your-model) instead. This warning will be raised to an exception in v4.41.
Non-default generation parameters: {'max_length': 4096}
Your generation config was originally created from the model config, but the model config has changed since then. Unless you pass the `generation_config` argument to this model's `generate` calls, they will revert to the legacy behavior where the base `generate` parameterization is loaded from the model config instead. To avoid this behavior and this warning, we recommend you to overwrite the generation config model attribute before calling the model's `save_pretrained`, preferably also removing any generation kwargs from the model

('./merged_model/tokenizer_config.json',
 './merged_model/special_tokens_map.json',
 './merged_model/tokenizer.model',
 './merged_model/added_tokens.json',
 './merged_model/tokenizer.json')

In [106]:
#save a merged model to huggingface
merged_model.push_to_hub("fined_tuned_llama2_for_finance_sentiment", use_temp_dir=True)
tokenizer.push_to_hub("fined_tuned_llama2_for_finance_sentiment", use_temp_dir=True)

Some non-default generation parameters are set in the model config. These should go into a GenerationConfig file (https://huggingface.co/docs/transformers/generation_strategies#save-a-custom-decoding-strategy-with-your-model) instead. This warning will be raised to an exception in v4.41.
Non-default generation parameters: {'max_length': 4096}


model-00003-of-00003.safetensors:   0%|          | 0.00/3.59G [00:00<?, ?B/s]

Upload 3 LFS files:   0%|          | 0/3 [00:00<?, ?it/s]

model-00002-of-00003.safetensors:   0%|          | 0.00/4.95G [00:00<?, ?B/s]

model-00001-of-00003.safetensors:   0%|          | 0.00/4.94G [00:00<?, ?B/s]

NameError: name 'tokenizer' is not defined

# Testing 

The following code will first predict the sentiment labels for the test set using the predict() function. Then, it will evaluate the model's perfomance on the test set using the evaluate() function. The result now should be impressive with an overall accuracy of over 0.8 and high accuracy, precision and recall for the single sentiment labels. The prediction of the neutral label can still be improved, yet it is impressive how much could be done with little data and some fine-tuning.

In [27]:
y_pred = predict(test, merged_model, tokenizer)
evaluate(y_true, y_pred)

100%|██████████| 900/900 [03:50<00:00,  3.91it/s]

Accuracy: 0.853
Accuracy for label 0: 0.913
Accuracy for label 1: 0.863
Accuracy for label 2: 0.783

Classification Report:
              precision    recall  f1-score   support

           0       0.95      0.91      0.93       300
           1       0.75      0.86      0.80       300
           2       0.88      0.78      0.83       300

    accuracy                           0.85       900
   macro avg       0.86      0.85      0.85       900
weighted avg       0.86      0.85      0.85       900


Confusion Matrix:
[[274  24   2]
 [ 11 259  30]
 [  3  62 235]]


The following code will create a Pandas DataFrame called evaluation containing the text,true labels, and predicted labels from the test set. This is expectially useful for understanding the errors that the fine-tuned model makes, and getting insights on how to improve the prompt.

In [29]:
evaluation = pd.DataFrame({'text': X_test["text"], 
                           'y_true':y_true, 
                           'y_pred': y_pred},
                         )
evaluation.to_csv("test_predictions.csv", index=False)

# Sentiment Analysis of Microsoft and Tesla News Articles

In [92]:
import pandas as pd

filename = "../input/microsoft-tesla-finance-news-articles2020-2024/tsla_articles.csv"

tsla = pd.read_csv(filename,
                 encoding="utf-8", encoding_errors="replace",index_col=0)

In [93]:
tsla

date                                               text
0     2021-01-27  The company benefited from a jump in sales of ...
1     2021-01-19  Rivian, which has raised another $2.65 billion...
2     2021-01-15  Traditional automakers have struggled to sell ...
3     2021-01-22  The pandemic dampened sales for all automakers...
4     2021-01-07  Mr. Musk’s net worth was $188.5 billion at 10:...
...          ...                                                ...
1409  2020-12-02  The president-elect’s economic team says that ...
1410  2020-12-28  Mike Strizki powers his house and cars with hy...
1411  2020-12-09  A panel of experts debate what post-Covid poli...
1412  2020-12-26  Investors of all stripes piled into stocks thi...
1413  2020-12-01  It will ask the S.E.C. to approve a new rule r...

[1414 rows x 2 columns]

In [94]:
def generate_sentiment_analysis_prompt(data_point):
    return f"""
            Analyze the sentiment of the news headline enclosed in square brackets, 
            determine if it is positive, neutral, or negative, and return the answer as 
            the corresponding sentiment label "positive" or "neutral" or "negative".

            [{data_point["text"]}] = """.strip()

In [95]:
article_input = pd.DataFrame(tsla.apply(generate_sentiment_analysis_prompt, axis=1), 
                       columns=["text"])

In [96]:
article_input

text
0     Analyze the sentiment of the news headline enc...
1     Analyze the sentiment of the news headline enc...
2     Analyze the sentiment of the news headline enc...
3     Analyze the sentiment of the news headline enc...
4     Analyze the sentiment of the news headline enc...
...                                                 ...
1409  Analyze the sentiment of the news headline enc...
1410  Analyze the sentiment of the news headline enc...
1411  Analyze the sentiment of the news headline enc...
1412  Analyze the sentiment of the news headline enc...
1413  Analyze the sentiment of the news headline enc...

[1414 rows x 1 columns]

In [97]:
def get_sentiment_score(test, model, tokenizer):
    y_pred = []
    for i in tqdm(range(len(article_input))):
        prompt = article_input.iloc[i]["text"]
        pipe = pipeline(task="text-generation", 
                        model=model, 
                        tokenizer=tokenizer, 
                        max_new_tokens = 1, 
                        temperature = 0.0,
                       )
        result = pipe(prompt)
        answer = result[0]['generated_text'].split("=")[-1]
        if "positive" in answer:
            y_pred.append("positive")
        elif "negative" in answer:
            y_pred.append("negative")
        elif "neutral" in answer:
            y_pred.append("neutral")
        else:
            y_pred.append("none")
    return y_pred

In [98]:
sentiment_labels = get_sentiment_score(test, merged_model, tokenizer)

100%|██████████| 1414/1414 [06:41<00:00,  3.52it/s]


In [99]:
labels_df = pd.DataFrame(sentiment_labels, columns=['label'])
tsla['label'] = labels_df
tsla.to_csv("tsla_with_sentiment_label.csv", index=False)

In [102]:
len(sentiment_labels)

1414

In [100]:
tsla

date                                               text     label
0     2021-01-27  The company benefited from a jump in sales of ...  negative
1     2021-01-19  Rivian, which has raised another $2.65 billion...   neutral
2     2021-01-15  Traditional automakers have struggled to sell ...   neutral
3     2021-01-22  The pandemic dampened sales for all automakers...  negative
4     2021-01-07  Mr. Musk’s net worth was $188.5 billion at 10:...  positive
...          ...                                                ...       ...
1409  2020-12-02  The president-elect’s economic team says that ...   neutral
1410  2020-12-28  Mike Strizki powers his house and cars with hy...   neutral
1411  2020-12-09  A panel of experts debate what post-Covid poli...   neutral
1412  2020-12-26  Investors of all stripes piled into stocks thi...   neutral
1413  2020-12-01  It will ask the S.E.C. to approve a new rule r...   neutral

[1414 rows x 3 columns]